### PR TITLE
Issue 5372: Maintain compatibility for older Credentials interface

### LINKED ISSUE #5372: Credentials interface was moved, which can break existing implementations
**Problem description**

First the background: PR #5339 recently moved `Credentials` interface from package `io.pravega.client.stream.impl` to `io.pravega.shared.security.auth`: 
https://github.com/pravega/pravega/pull/5339/files#diff-3c7869f3942a642efca0109a6c26ae6b63713c5e6debf89ad65322474bbc802c

The rationale for the change of this external-facing interface is to:
* Move it to shared:security module so that the interface can be implemented without the need to add a dependency on Pravega client . 
* Move it to a more appropriate package - together with other major security classes, so that developers can locate it more easily. 

However, this change can break any custom implementation of the interface at a compilation-level. Since this is an external-facing interface, a more prudent approach is to deprecate it first, and remove it later. 

**Problem location**

**Suggestions for an improvement**

Allow implementations of the old interface to work, along with implementations of the relocated interface.

### PR DESCRIPTION
**Pull request by ravisharda**
**Change log description**  

Support `io.pravega.client.stream.impl.Credentials` interface along with the relocated `io.pravega.shared.security.auth.Credentials` interface, for compatibility.

**Purpose of the change**  
Fixes #5372. 

**What the code does**  
* Brings back the older `io.pravega.client.stream.impl.Credentials` interface. It now implements `io.pravega.shared.security.auth.Credentials` 

**How to verify it**  
All unit and integration tests must pass.

### CODE REVIEW

**Review by shrids — COMMENTED** [review]
Changes look good apart from one comment.

File: client/src/test/java/io/pravega/client/CredentialsExtractorTest.java [hunk 1]
```diff
@@ -190,6 +190,43 @@ public void testLoadsCredentialsObjOfARegisteredTypeFromEnvVariablesIfLoadDynami
...
+        Credentials credentials = clientConfig.getCredentials();
+
+        assertNotNull("Credentials is null", credentials);
+        assertNotNull(LegacyCredentials1.class.getName(), credentials.getClass());
+        assertEquals("Expected a different authentication type", LegacyCredentials1.AUTHENTICATION_METHOD,
+                credentials.getAuthenticationType());
+    }
+
+    @Test
+    public void testLoadsLegacyCredentialsUsingOldInterfacePackage() {
```
[the remark this answers is not in the document]
Comment (ravisharda) [reply client/src/test/java/io/pravega/client/CredentialsExtractorTest.java:213]: The difference in the tests is in how they are setup to load the legacy `Credentials` class. 

One is setup using for service loading using the implemented interface: 

https://github.com/pravega/pravega/blob/7bbc7d9abac42d6d125b961406cd3f2767b0709a/client/src/test/resources/META-INF/services/io.pravega.shared.security.auth.Credentials

This simulates the scenario where the existing implementations of the `io.pravega.client.stream.impl.Credentials` will continue to work if they are specified using the new interface's service file: `io.pravega.shared.security.auth.Credentials`. 

The other is setup using the same interface: 

https://github.com/pravega/pravega/blob/7bbc7d9abac42d6d125b961406cd3f2767b0709a/client/src/test/resources/META-INF/services/io.pravega.client.stream.impl.Credentials

This simulates the scenario where the existing implementations of the `io.pravega.client.stream.impl.Credentials` will also work if they are specified using the old interface's service file: `io.pravega.client.stream.impl.Credentials`. So, they will work without any code or configuration changes. 

Hope that clarifies.

[the remark this answers is not in the document]
Comment (shrids) [reply client/src/test/java/io/pravega/client/CredentialsExtractorTest.java:213]: Thanks for the clarification. 👍

**Review by shrids — APPROVED** [review]

**Review by fpj — APPROVED** [review]
Could you please create an issue for us to remove the deprecated interface in 0.10?

**Comment by ravisharda** [comment]
> Could you please create an issue for us to remove the deprecated interface in 0.10?

Sure, here's the issue: #5383.

**Pull request merged**